flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TriggerManager.java

```diff
@@ -0,0 +1,296 @@
+                    "triggerManagerAccumulatedChange", TypeInformation.of(TableChange.class)));
+    this.lastTriggerTimes =
+        getRuntimeContext()
+            .getListState(new ListStateDescriptor<>("triggerManagerLastTriggerTime", Types.LONG));
+
+    tableLoader.open();
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext context) throws Exception {
+    // Do nothing
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) {
+    LOG.info("Initializing state restored: {}", context.isRestored());
+    this.lock = lockFactory.createLock();
+    this.recoveryLock = lockFactory.createRecoveryLock();
+    if (context.isRestored()) {
+      isCleanUp = true;
+    }
+  }
+
+  @Override
+  public void processElement(TableChange change, Context ctx, Collector<Trigger> out)
+      throws Exception {
+    init(out, ctx.timerService());
+
+    long current = ctx.timerService().currentProcessingTime();
+    Long nextTime = nextEvaluationTime.value();
+
+    // Add the new changes to the already accumulated ones
+    List<TableChange> accumulated = Lists.newArrayList(accumulatedChanges.get());
+    accumulated.forEach(tableChange -> tableChange.merge(change));
+    accumulatedChanges.update(accumulated);
+
+    if (nextTime == null) {
+      checkAndFire(current, ctx.timerService(), out);
+    } else {
+      LOG.info(
+          "Trigger manager rate limiter triggered current: {}, next: {}, accumulated changes: {}",
+          current,
+          nextTime,
+          accumulated);
+      rateLimiterTriggeredCounter.inc();
+    }
+  }
+
+  @Override
+  public void onTimer(long timestamp, OnTimerContext ctx, Collector<Trigger> out) throws Exception {
+    init(out, ctx.timerService());
+    nextEvaluationTime.clear();
+    checkAndFire(ctx.timerService().currentProcessingTime(), ctx.timerService(), out);
+  }
+
+  private void checkAndFire(long current, TimerService timerService, Collector<Trigger> out)
+      throws Exception {
+    if (isCleanUp) {
+      if (recoveryLock.isHeld()) {
+        LOG.debug("The cleanup lock is still held at {}", current);
+        schedule(timerService, current + lockCheckDelayMs);
+        return;
+      } else {
+        LOG.info("The cleanup is finished at {}", current);
+        isCleanUp = false;
+      }
+    }
+
+    List<TableChange> changes = Lists.newArrayList(accumulatedChanges.get());
+    List<Long> times = Lists.newArrayList(lastTriggerTimes.get());
+    Integer taskToStart = nextTrigger(evaluators, changes, times, current, startsFrom);
+    if (taskToStart == null) {
+      // Nothing to execute
+      if (startsFrom == 0) {
+        nothingToTriggerCounter.inc();
+        LOG.debug("Nothing to execute at {} for collected: {}", current, changes);
+      }
+
+      startsFrom = 0;
+      return;
+    }
+
+    if (lock.tryLock()) {
+      TableChange change = changes.get(taskToStart);
+      SerializableTable table =
```

Comment:
for each `Trigger` record, we are loading the table from catalog. why do we let downstream operator load the table directly? i.e. downstream operator gets the `TableLoader` and we avoid including `SerializableTable` to the `Trigger` object?

Reply:
Most of the tasks need Table anyways. If we do not do it here, we will have a `SnapshotTable` operator in:
- RewriteDataFiles
- RewriteManifestFiles
- DeleteOrphanFiles

We do not need the table for:
- ExpireSnapshots - as we need to load the table anyways

I'm open to both solutions. I'm opted for reading the table here to simplify the specific Maintenance Tasks later.